packages/nimble-blazor/NimbleBlazor/NimbleInputBase.cs

```diff
@@ -233,48 +233,40 @@ private void UpdateAdditionalValidationAttributes()
         }
 
         var hasAriaInvalidAttribute = AdditionalAttributes != null && AdditionalAttributes.ContainsKey("aria-invalid");
-        if (EditContext.GetValidationMessages(FieldIdentifier).Any())
-        {
-            if (hasAriaInvalidAttribute)
-            {
-                // Do not overwrite the attribute value
-                return;
-            }
+        var validationMessages = EditContext.GetValidationMessages(FieldIdentifier).ToArray();
 
+        if (validationMessages.Any())
+        {
             if (ConvertToDictionary(AdditionalAttributes, out var additionalAttributes))
             {
                 AdditionalAttributes = additionalAttributes;
             }
 
             // To make the `Input` components accessible by default
-            // we will automatically render the `aria-invalid` attribute when the validation fails
+            // we will automatically render the `aria-invalid`, `error-visible` and `error-text` attributes when the validation fails
             additionalAttributes["aria-invalid"] = true;
+            additionalAttributes["error-visible"] = true;
+            additionalAttributes["error-text"] = validationMessages.First();
```

Comment:
To begin with, we ultimately [would like to move away from the `AdditionalAttributes` pattern](https://github.com/ni/nimble/issues/751) to promote stronger typing of known attributes that we want clearly exposed on our APIs (though this could be a problematic goal, given how extensively this could break people).

Additionally, introducing knowledge of the `error-visible` and `error-text` attributes into this class seems like a subversion of the existing APIs that are present (i.e. `NimbleInputBase` does not own the `ErrorText` or `ErrorVisible` properties). It's possible that this is potentially how the API should evolve, where  `NimbleInputBase` does indeed expose those properties, but I haven't fully wrapped my head around the 
implications/benefits/drawbacks of this yet.

At the moment, I don't have a strong sense for the right way to provide the capabilities you are requesting. The approach that I'm toying with in my head is something like the following:
- Create a new intermediate abstract Component class `ErrorInput`, which hosts the `ErrorText` and `ErrorVisible` properties.
    - Additionally, this class would have public read-only properties (with protected setters) `ActualErrorInput` and `ActualErrorVisible`. Updates to `ErrorText` and `ErrorVisible` would update their `Actual<>` counterparts. The reasoning behind a pattern like this is to prevent stomping on the user-settable properties of the components. The various concrete `ErrorInput` components (like `NimbleTextField`) would then bind their attributes to the `Actual<>` properties.
- Make `OnValidateStateChanged` on `NimbleInputBase` `protected virtual`.
- Override `OnValidteStateChanged` on `ErrorInput` and set the `Actual<>` properties as needed with the validation information available on the `EditContent`.

I'm also unsure if we should be handling the `aria-invalid` attribute at this level, but I think that is ultimately orthogonal to the capabilities we're trying to add here.

@rajsite and @msmithNI please chime in with your own thoughts as well.

Reply:
I think my concerns are:
- +1 that we want to move away from AdditionalAttributes to provide stronger typing
- It's not clear if the user intentionally specifies ErrorText and ErrorVisible how that interacts with the validation results. I think this is testable in the current infrastructure, i.e. testing validation behavior and emitted attributes.
- My understanding is that Blazor lets you opt-in visualizing validation state. You have to manually add ValidationMessage or ValidationSummary components to do the visualization.

   We define the visual for validation sure (via error-text and error-visible), but I think we should have a way to opt-in or opt-out of the presentation for a control in an EditForm context. We have talked about it being frustrating that just because we are in an edit context we are forced to define a [value expression](https://github.com/ni/nimble/blob/b90b5fa475b233425c49275c6011c96d8c2b1013/packages/nimble-blazor/NimbleBlazor/NimbleInputBase.cs#L190). 

Maybe we say that by default most users want the automatic validation and make it opt-out?

```html
<NimbleTextField CustomError="True" @bind-Value="exampleModel.MyValue" ErrorText="My custom message" @bind-ErrorVisible="exampleModel.ErrorVisible">
```

Or to make it opt-in which is more like Blazor's ValidationMessage pattern, let you select the validation context (i.e. something like the `For` property):

```html
<NimbleTextField @bind-Value="exampleModel.MyValue" Error-For="@(() => exampleModel.MyValue)">
```

That gives you a lot of control, the error state of the text field doesn't have to correspond only to the text field value, but maybe that is not a useful degree of freedom.

Reply:
I will really need this in H2, but wanted to go ahead and make sure this will be completed until then. I can still use the `ValidationMessage` component until this gets completed.